blackbird_python/blackbird/listener.py

```diff
@@ -342,16 +353,49 @@ def exitArrayvar(self, ctx: blackbirdParser.ArrayvarContext):
                 )
             )
 
-        if shape is not None:
+        # if the array is defined as a single parameter, then split it up into the correct shape and
+        # rename the parameters to 'parametername_i_j' for indices i and j in the array
+        if final_value.size == 0 and len(parameters) == 1:
+            if not ctx.shape():
+                line = ctx.start.line
+                col = ctx.start.column
+                raise BlackbirdSyntaxError(
+                    "Blackbird SyntaxError (line {}:{}): Array template var {} has no shape defined.".format(
+                        line, col, name
+                    )
+                )
+
+            final_value = []
+            for i in range(shape[0]):
+                final_value.append([])
+                for j in range(shape[1]):
+                    param = sym.Symbol(parameters[0][1].name + "_{}_{}".format(i, j))
+                    final_value[-1].append(param)
+            final_value = np.array(final_value)
+
+            _PARAMS.extend(final_value.flatten())
+            _PARAMS.remove(parameters[0][1])
+        else:
+            # if any elements are parameters, then change array type to object and
+            # re-insert parameters into array at the correct postion
+            if parameters:
+                final_value = final_value.astype(object)
+                for p in parameters:
+                    final_value = np.insert(final_value, p[0], p[1])
+
+            # reshape the array into the correct shape and check with declared shape
+            final_value = final_value.reshape(array_rows, -1)
             actual_shape = final_value.shape
-            if actual_shape != shape:
+            if shape and actual_shape != shape:
                 line = ctx.start.line
                 col = ctx.start.column
                 raise BlackbirdSyntaxError(
                     "Blackbird SyntaxError (line {}:{}): Array var {} has declared shape {} "
                     "but actual shape {}".format(line, col, name, shape, actual_shape)
                 )
 
+        if self._program._type["name"] == "tdm" and name in ["BS", "M", "R"]:
+            _PARAMS.append(name)
```

Comment:
It's not very clear what is happening here, so worth adding a comment.

Also, how come this branch is only taken if the array variable is named `BS`, `M`, or `R`? With the move to `p\d` naming, is this a candidate for removal?